discoverd2/server/http.go

```diff
@@ -0,0 +1,217 @@
+package server
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+	"strings"
+
+	"github.com/flynn/flynn/discoverd2/client"
+	"github.com/flynn/flynn/pkg/httphelper"
+	"github.com/flynn/flynn/pkg/sse"
+	"github.com/flynn/flynn/pkg/stream"
+	"github.com/julienschmidt/httprouter"
+)
+
+type Datastore interface {
+	// Typically implemented by a Backend
+	AddService(service string) error
+	RemoveService(service string) error
+	AddInstance(service string, inst *discoverd.Instance) error
+	RemoveInstance(service, id string) error
+
+	// Typically implemented by State
+	Get(service string) []*discoverd.Instance
+	GetLeader(service string) *discoverd.Instance
+	Subscribe(service string, sendCurrent bool, kinds discoverd.EventKind, ch chan *discoverd.Event) stream.Stream
+}
+
+type basicDatastore struct {
+	*State
+	Backend
+}
+
+func (d basicDatastore) AddService(service string) error {
+	return d.Backend.AddService(service)
+}
+
+func (d basicDatastore) RemoveService(service string) error {
+	return d.Backend.RemoveService(service)
+}
+
+func (d basicDatastore) AddInstance(service string, inst *discoverd.Instance) error {
+	return d.Backend.AddInstance(service, inst)
+}
+
+func (d basicDatastore) RemoveInstance(service, id string) error {
+	return d.Backend.RemoveInstance(service, id)
+}
+
+func NewBasicDatastore(state *State, backend Backend) Datastore {
+	return &basicDatastore{state, backend}
+}
+
+func NewHTTPHandler(ds Datastore) http.Handler {
+	router := httprouter.New()
+
+	api := &httpAPI{
+		Store: ds,
+	}
+
+	router.PUT("/services/:service", api.AddService)
+	router.DELETE("/services/:service", api.RemoveService)
+	router.GET("/services/:service", api.GetServiceStream)
+
+	router.PUT("/services/:service/instances/:instance_id", api.AddInstance)
+	router.DELETE("/services/:service/instances/:instance_id", api.RemoveInstance)
+	router.GET("/services/:service/instances", api.GetInstances)
+
+	router.GET("/services/:service/leader", api.GetLeader)
+
+	return router
+}
+
+type httpAPI struct {
+	Store Datastore
+}
+
+func jsonError(w http.ResponseWriter, code int, err error) {
+	httphelper.JSON(w, code, struct {
+		Error string `json:"error"`
+	}{err.Error()})
```

Comment:
Might make sense to lift into httphelper?

Should we consider additional error enum strings that are well defined and exported, so the remote can easily check for them?  Maybe the error messages would be enough, but I always feel a little uncomfortable checking a string that isn't clearly marked as not-for-humans/do-not-refactor-phrasing.